mock.go

```diff
@@ -909,3 +913,31 @@ func assignRecords(m map[string]interface{}, record map[string]interface{}) erro
 
 	return nil
 }
+
+type mockContextKey string
+
+var partialFailureContextKey mockContextKey = "partial_failure"
+
+type ShouldFailFunc func(opIdx, opCount int) error
```

Comment:
I think some docs would be helpful on the following:
 * what `opIdx` and `opCount` mean
 * gotchas e.g. `idx` is 0 indexed but `opCount` isn't so `opCount - 1` will give the user the final op not the second to last one.
 * What purposes have we designed `PartialFailureContext` to satisfy?

Reply:
Should we add in example test cases? (or at least test that this behaves as we expect)